src/netaddress.cpp

```diff
@@ -287,17 +285,12 @@ bool CNetAddr::IsValid() const
     if (IsInternal())
         return false;
 
-    if (IsIPv4())
-    {
-        // INADDR_NONE
-        uint32_t ipNone = INADDR_NONE;
-        if (memcmp(ip+12, &ipNone, 4) == 0)
-            return false;
-
-        // 0
-        ipNone = 0;
-        if (memcmp(ip+12, &ipNone, 4) == 0)
-            return false;
+    if (IsIPv4()) {
+        for (uint32_t a : {htonl(INADDR_ANY), htonl(INADDR_NONE)}) {
+            if (memcmp(m_addr.data(), &a, sizeof(a)) == 0) {
+                return false;
+            }
+        }
```

Comment:
The opposite seems more straightforward:
```C++
if (IsIPv4()) {
    uint32_t a;
    memcpy(a, m_addr.data(), sizeof(a));
    if (a == htonl(INADDR_ANY) || a == htonl(INADDR_NONE)) return false;
}
```
@sipa might yell at me but it might also work to do `uint32_t a = *(uint32_t*)m_addr.data();` to skip the `memcpy` line.

Reply:
Those constants are defined like this in the system headers:

```cpp
typedef uint32_t in_addr_t;
...
#define	INADDR_ANY		((in_addr_t) 0x00000000)
#define	INADDR_NONE		((in_addr_t) 0xffffffff)
...
#define INADDR_ALLHOSTS_GROUP	((in_addr_t) 0xe0000001) /* 224.0.0.1 */
```

Of course any discussion about the byte order of `INADDR_ANY` is meaningless, but nevertheless - we can assume that all `INADDR_*` are in the same byte order. From the above we see that 224.0.0.1 is represented as the integer `3758096385` or `0xe0000001` which on LE systems is stored in memory as `01 00 00 e0` and on BE systems as `e0 00 00 01`.

This means that whenever we want to compare those constants to something we know is in BE byte order (like `m_addr`) we have to apply `htonl()` to them. This is how it is done in the [L](http://tomoyo.osdn.jp/cgi-bin/lxr/source/net/bridge/br_multicast.c?a=sparc64#L263)inu[x](http://tomoyo.osdn.jp/cgi-bin/lxr/source/net/ipv4/igmp.c?a=sparc64#L1542) and [FreeBSD](https://github.com/freebsd/freebsd/blob/300d08e9968a2e38e163d3c3479524a587fe8577/sys/netinet/in.h#L658) system headers.

So, leaving this as is at this iteration.

Reply:
You're right.

I was confused by the fact that `struct in_addr`'s `s_addr` field stores IP addresses in BE order, but the INADDR_* constants are in native order (and you are supposed to use `htonl` & friends to convert between them; see the section on `sin_addr` in https://man7.org/linux/man-pages/man7/ip.7.html).

Arguably you should be using `ntohl` instead of `htonl` (as these constants are in native order, and we want them in BE order) - but on every realistic platform these two are equivalent.

If you want to avoid the winsock/inet header dependency you can also use WriteBE32 to convert them to bytes.

Reply:
as @sipa said 
```C++
uint32_t a = ReadBE32(m_data.data());
if (a == INADDR_ANY || a == INADDR_ALL) ...
```
is so much more readable :)

Reply:
Oops, yes, I guess this confused @kallewoof too (on IRC). Specifially, this is UB:

```c++
uint32_t a = *(uint32_t*)m_addr.data();
```

as it's accessing a byte array through an incompatible type.